clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -0,0 +1,273 @@
+        this.groupState = groupState;
+        this.subscriptionState = subscriptionState;
+    }
+
+    // Visible for testing
+    CommitRequestManager(
+            final Time time,
+            final LogContext logContext,
+            final SubscriptionState subscriptionState,
+            final ConsumerConfig config,
+            final CoordinatorRequestManager coordinatorRequestManager,
+            final GroupStateManager groupState,
+            final AutoCommitState autoCommitState) {
+        this.log = logContext.logger(getClass());
+        this.subscriptionState = subscriptionState;
+        this.coordinatorRequestManager = Optional.ofNullable(coordinatorRequestManager);
+        this.groupState = groupState;
+        this.autoCommitState = Optional.ofNullable(autoCommitState);
+        this.stagedCommits = new LinkedList<>();
+    }
+
+    /**
+     * Poll for the commit request if there's any. The function will also try to autocommit, if enabled.
+     *
+     * @param currentTimeMs
+     * @return
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (!coordinatorRequestManager.isPresent()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>());
+        }
+
+        maybeAutoCommit(currentTimeMs);
+
+        if (stagedCommits.isEmpty()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>());
+        }
+
+        List<NetworkClientDelegate.UnsentRequest> unsentCommitRequests =
+                stagedCommits.stream().map(StagedCommit::toUnsentRequest).collect(Collectors.toList());
+        return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.unmodifiableList(unsentCommitRequests));
+    }
+
+    public RequestFuture<ClientResponse> add(final Map<TopicPartition, OffsetAndMetadata> offsets,
+                                             final Optional<OffsetCommitCallback> callback) {
+        OffsetCommitRequestHandler handler = new OffsetCommitRequestHandler(offsets, callback.orElse(null));
+        this.add(offsets, handler);
+        return handler.future();
+    }
+
+    public RequestFuture<ClientResponse> add(final Map<TopicPartition, OffsetAndMetadata> offsets,
+                                             final NetworkClientDelegate.AbstractRequestFutureCompletionHandler handler) {
+        this.stagedCommits.add(new StagedCommit(
+                offsets,
+                handler,
+                groupState.groupId,
+                groupState.groupInstanceId.orElse(null),
+                groupState.generation));
+        return handler.future();
+    }
+
+    // TODO: I think we might need to return the callback because we will need to send autocommit and wait for its
+    //  completion before joinGroup. Alternatively, we could try to change the join group state upon success or
+    //  failure.
+    private void maybeAutoCommit(final long currentTimeMs) {
```

Comment:
this method is still WIP: because we need to finalize on the subscriptions state changes. my initial thought was using the consumed and fetch positions to determine if the previous fetch has been consumed or not.